lib/puppet/parser/ast/leaf.rb

```diff
@@ -27,7 +27,7 @@ def initialize(value: nil, **options)
 #
 class Puppet::Parser::AST::HostName < Puppet::Parser::AST::Leaf
   def initialize(hash)
-    super
+    super(**hash)
```

Comment:
it seems odd to me that `HostName` takes a hash, but its parent `Leaf` takes keyword arguments. Maybe better to make it explicit so `HostName` has the same signature as its parent `def initialize(file: nil, line: nil, pos: nil)`? I'm assuming this wasn't noticed in 4139dd0c96e because the tests passed and there weren't any warnings before 2.7?

Reply:
`HostName` inherits from `Leaf`, which inherits from `Puppet::Parser::AST`

```
class Puppet::Parser::AST::Leaf < Puppet::Parser::AST
  def initialize(value: nil, **options)
    @value = value
    super(**options)
  end
```

```
class Puppet::Parser::AST
  def initialize(file: nil, line: nil, pos: nil)
    @file = file
    @line = line
    @pos = pos
  end
```

from my understanding, passing in the hash as `**hash` here, it will behave as keyword arguments? So, for the call to super in `Leaf`, if `options` contains `file`, `line`, and/or `pos`, then those will be populated. Otherwise, anything additional will be disregarded. Here, if our `hash` contains `value`, it'll be used to populate `@value`, and everything else in the has will be sent to the `AST` initializer.

I'm happy to update this to make it more consistent, but I feel like we would also want to change the initializer for `Leaf` in addition to this one. So, final product would be
```
class Puppet::Parser::AST::HostName < Puppet::Parser::AST::Leaf
  def initialize(value: nil, file: nil, line: nil, pos: nil)
    super(value: value, file: file, line: line, pos: pos)

    # Note that this is an AST::Regex, not a Regexp
    unless @value.is_a?(Regex)
      @value = @value.to_s.downcase
      if @value =~ /[^-\w.]/
        raise Puppet::DevError, _("'%{value}' is not a valid hostname") % { value: @value }
      end
    end
  end
end

class Puppet::Parser::AST::Leaf < Puppet::Parser::AST
  def initialize(value: nil, file: nil, line: nil, pos: nil)
    @value = value
    super(file: file, line: line, pos: pos)
  en
end

class Puppet::Parser::AST
  def initialize(file: nil, line: nil, pos: nil)
    @file = file
    @line = line
    @pos = pos
  end
end
```

but that doesn't quite match up with what you were saying. I definitely could have overlooked something. Thoughts?